### PR TITLE
Convert body text to UTF-8 before encrypting for signature

### PR DESCRIPTION
**Pull request by benmccormick**
Hi!  

At Windsor Circle we have started testing the Signature based authentication and found that our (python-based) encryption didn't match the webhooks signature for JSON text that contained unicode characters.  This is due to JavaScripts strange default handling of unicode where each code point is treated as a separate character. If we treat all strings as UTF-8, this solves this problem for strings containing unicode, without breaking anything for non-unicode strings.

An example bad request:

```
'{"anonymousId":"f9ad2ff3-77e8-47d7-a3fe-339e9bf7052c","channel":"client","context":{"ip":"66.57.36.207","library":{"name":"analytics.js","version":"2.10.0"},"page":{"path":"/products/212781","referrer":"https://smitham-group8975.myshopify.com/collections/frontpage/products/public-key-holistic-budgetary-management","search":"","title":"Acoustic Research ARRX18G Xsight Touch Universal Remote Control – BDHE","url":"http://smitham-group8975.myshopify.com/products/212781"},"userAgent":"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/45.0.2454.85 Safari/537.36"},"event":"Viewed Product","integrations":{},"messageId":"ajs-0c37552eac671297076879cffcb6a772","originalTimestamp":"2015-09-15T18:51:19.900Z","projectId":"jGdE3nzNlv","properties":{"available":true,"collections":["Frontpage"],"id":"91563458","name":"Acoustic Research ARRX18G Xsight Touch Universal Remote Control","options":["Title"],"optionsCount":1,"price":"24999","priceMax":"24999","priceMin":"24999","priceVaries":"false","tag":[""],"type":"Home Theater > Power Amplifiers> Headphone Amplifiers","url":"/products/212781","vendor":"Acoustic Research"},"receivedAt":"2015-09-15T18:51:19.937Z","sentAt":"2015-09-15T18:51:19.902Z","timestamp":"2015-09-15T18:51:19.935Z","type":"track","userId":null,"version":2}'
```

This produces `b64c8bc284dbcc6962d906188156cb5959437aba` when run through the supplied JS code, whereas our python code produces `e143ca1cfb062fb0aec746ecc004e6422b570d38` using our key (obviously the exact values will vary dependent on key).  Values match up fine when unicode characters are not present

This is because of the unicode character in `Acoustic Research ARRX18G Xsight Touch Universal Remote Control – BDHE`.  When that dash at the end is removed, the digests match. 

The python code we're using is below

```python
import hmac
import hashlib
event = '<event string>'
key ='<our secret>'

encryption = hmac.new(key, event, hashlib.sha1)
print(encryption.hexdigest())
```

More on JavaScript weirdness with unicode:

https://mathiasbynens.be/notes/javascript-encoding
https://mathiasbynens.be/notes/javascript-unicode

### CODE REVIEW

**Comment by hankim813** [comment]
@ndhoule @f2prateek @bbeaird can you guys review this?

**Comment by benmccormick** [comment]
Just to help a bit more with this:  I'm verifying the behavior using the following test cases:

### event.js

```
#! /usr/bin/env node
var crypto = require('crypto');
var readline = require('readline');
var rl = readline.createInterface({
  input: process.stdin,
});
var secret = 'secret';
var useUTF8 = process.argv[2] === '--useUTF8';
rl.on('line', function(line){
   if(useUTF8) {
      line = unescape(encodeURIComponent(line));
   }
   var digest = crypto.createHmac('sha1', secret).update(line).digest('hex');
   console.log(digest);
});
```

### event.py

```
#!/usr/bin/env python
import sys
import hmac
import hashlib
secret = 'secret'

for line in sys.stdin:
    string = line.strip()
    encryption = hmac.new(secret, string, hashlib.sha1)
    print(encryption.hexdigest())
```

Here's the results of running those scripts against unicode and non-unicode test strings with and without the additional JavaScript code:

1__uiworkspace___users_ben_dropbox_blog_drafts__tmux__and___code_windsor_scratch_browse-events_event_js_and_profiles_and_javascript_-_how_to_pass_command_line_arguments_to_node_js__-_stack_overflow

Basically, the encoding change has no effect on ascii strings, but it will cause the unicode strings to be encoded as UTF-8.  That will then be read normally by any language that treats unicode strings as UTF-8 by default, and serves as a sane format for other languages to encode in, since JavaScript's default implementation ([sort of a hybrid between UCS-2 and UTF-16](https://mathiasbynens.be/notes/javascript-encoding)) is not something that is easy to target from another language (still haven't figured out how to do it in Python)

**Comment by benmccormick** [comment]
I just dug in a bit more and discovered that crypto's hmac.update function accepts an encoding option.  Since thats clearer and less hacky looking, I've updated the PR to use it.

**Comment by f2prateek** [comment]
LGTM!

**Comment by f2prateek** [comment]
Thanks for digging into this @benmccormick .

**Comment by benmccormick** [comment]
@f2prateek Awesome!  Thanks for the quick merge :clap:
